pkg/envoy/server.go

```diff
@@ -495,33 +495,43 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, labelsMap identity.I
+func getNetworkPolicy(ipv6 addressing.CiliumIPv6, ipv4 addressing.CiliumIPv4, id identity.NumericIdentity, policy *policy.L4Policy,
+	labelsMap identity.IdentityCache, deniedIngressIdentities, deniedEgressIdentities map[identity.NumericIdentity]bool) (ipv6Policy, ipv4Policy *cilium.NetworkPolicy) {
 	return &cilium.NetworkPolicy{
-		Policy:                 uint64(id),
-		IngressPerPortPolicies: getDirectionNetworkPolicy(policy.Ingress, labelsMap, deniedIngressIdentities),
-		EgressPerPortPolicies:  getDirectionNetworkPolicy(policy.Egress, labelsMap, deniedEgressIdentities),
-	}
+			Name:                   ipv6.String(),
+			Identity:               uint64(id),
+			IngressPerPortPolicies: getDirectionNetworkPolicy(policy.Ingress, labelsMap, deniedIngressIdentities),
+			EgressPerPortPolicies:  getDirectionNetworkPolicy(policy.Egress, labelsMap, deniedEgressIdentities),
+		}, &cilium.NetworkPolicy{
+			Name:                   ipv4.String(),
+			Identity:               uint64(id),
+			IngressPerPortPolicies: getDirectionNetworkPolicy(policy.Ingress, labelsMap, deniedIngressIdentities),
+			EgressPerPortPolicies:  getDirectionNetworkPolicy(policy.Egress, labelsMap, deniedEgressIdentities),
+		}
 }
 
 // UpdateNetworkPolicy adds or updates a network policy in the set of published
 // to L7 proxies.
-func UpdateNetworkPolicy(id identity.NumericIdentity, policy *policy.L4Policy,
+func UpdateNetworkPolicy(ipv6 addressing.CiliumIPv6, ipv4 addressing.CiliumIPv4, id identity.NumericIdentity, policy *policy.L4Policy,
 	labelsMap identity.IdentityCache, deniedIngressIdentities, deniedEgressIdentities map[identity.NumericIdentity]bool) error {
-	networkPolicy := getNetworkPolicy(id, policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities)
-	err := networkPolicy.Validate()
+	v6Policy, v4Policy := getNetworkPolicy(ipv6, ipv4, id, policy, labelsMap, deniedIngressIdentities, deniedEgressIdentities)
+	err := v6Policy.Validate()
+	if err != nil {
+		return fmt.Errorf("error validating generated IPv6 NetworkPolicy: %s", err)
+	}
+	err = v4Policy.Validate()
 	if err != nil {
-		return fmt.Errorf("error validating generated NetworkPolicy: %s", err)
+		return fmt.Errorf("error validating generated IPv4 NetworkPolicy: %s", err)
 	}
 
-	name := strconv.FormatUint(uint64(id), 10)
-	NetworkPolicyCache.Upsert(NetworkPolicyTypeURL, name, networkPolicy, false)
+	NetworkPolicyCache.Upsert(NetworkPolicyTypeURL, v6Policy.Name, v6Policy, false)
+	NetworkPolicyCache.Upsert(NetworkPolicyTypeURL, v4Policy.Name, v4Policy, false)
 	return nil
 }
 
 // RemoveNetworkPolicy removes a network policy from the set published to L7
 // proxies.
-func RemoveNetworkPolicy(id identity.NumericIdentity) {
-	name := strconv.FormatUint(uint64(id), 10)
-	NetworkPolicyCache.Delete(NetworkPolicyTypeURL, name, false)
+func RemoveNetworkPolicy(ipv6 addressing.CiliumIPv6, ipv4 addressing.CiliumIPv4) {
+	NetworkPolicyCache.Delete(NetworkPolicyTypeURL, ipv6.String(), false)
```

Comment:
Check whether `ipv6` is `nil`, and don't delete in this case.
Same for `ipv4` below.